pallets/handles/src/handles_signed_extension.rs

```diff
@@ -0,0 +1,130 @@
+use core::marker::PhantomData;
+use frame_support::{
+	dispatch::{DispatchInfo, Dispatchable},
+	pallet_prelude::ValidTransaction,
+	traits::IsSubType,
+};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, SignedExtension},
+	transaction_validity::{InvalidTransaction, TransactionValidity, TransactionValidityError},
+	DispatchError,
+};
+/// The SignedExtension trait is implemented on CheckFreeExtrinsicUse to validate the request. The
+/// purpose of this is to ensure that the retire_handle extrinsic cannot be
+/// repeatedly called to flood the network.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct HandlesSignedExtension<T: Config + Send + Sync>(PhantomData<T>);
+
+impl<T: Config + Send + Sync> HandlesSignedExtension<T> {
+	/// Create new `SignedExtension`.
+	pub fn new() -> Self {
+		Self(sp_std::marker::PhantomData)
+	}
+}
+
+impl<T: Config + Send + Sync> sp_std::fmt::Debug for HandlesSignedExtension<T> {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "HandlesSignedExtension<{:?}>", self.0)
+	}
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+/// Validator trait to be used in validating Handles requests from a Signed Extension.
+pub trait HandlesValidate<C: Config> {
+	/// Validates calls to the retire_handle() extrinsic
+	fn validate_retire_handle(delegator_key: &C::AccountId) -> TransactionValidity;
+}
+
+impl<T: Config> HandlesValidate<T> for Pallet<T> {
+	/// Validates the following criteria for the retire_handle() extrinsic:
+	///
+	/// * The delegator must already have a MSA id
+	/// * The MSA must already have a handle associated with it
+	///
+	/// Returns a `ValidTransaction` or wrapped [`pallet::Error`]
+	///
+	/// # Errors (as u8 wrapped by `InvalidTransaction::Custom`)
+	/// * [`Error::InvalidMessageSourceAccount`]
+	/// * [`Error::MSAHandleDoesNotExist`]
+	fn validate_retire_handle(delegator_key: &T::AccountId) -> TransactionValidity {
+		const TAG_PREFIX: &str = "HandlesRetireHandle";
+
+		// Validation: The delegator must already have a MSA id
+		let delegator_msa_id = T::MsaInfoProvider::ensure_valid_msa_key(&delegator_key)
+			.map_err(|e| map_dispatch_error(e))?;
+		// Validation: The MSA must already have a handle associated with it
+		_ = MSAIdToDisplayName::<T>::try_get(delegator_msa_id)
+			.map_err(|_| map_dispatch_error(DispatchError::CannotLookup))?;
+		return ValidTransaction::with_tag_prefix(TAG_PREFIX).build()
+	}
+}
+
+/// Map a module DispatchError to an InvalidTransaction::Custom error
+pub fn map_dispatch_error(err: DispatchError) -> InvalidTransaction {
+	InvalidTransaction::Custom(match err {
+		DispatchError::Module(module_err) =>
+			<u32 as Decode>::decode(&mut module_err.error.as_slice())
+				.unwrap_or_default()
+				.try_into()
+				.unwrap_or_default(),
+		_ => 255u8,
```

Comment:
can you please explain what 255 is?

Reply:
`InvalidTransaction::Custom` takes a `u8`. If the `DispatchError` is not a `DispatchError::Module` from which we can extract a numeric code, then we just send the max u8 value (255) to indicate it's an unknown error. It's a little bit hacky, but there's no really good way to map a DispatchError to an InvalidTransaction error.

Reply:
I noticed that Substrate uses zero. Why did we decide to use 255?

Reply:
zero could be a valid error enum value from the pallet. 255 is highly unlikely to be.